webtau-junit5/src/main/java/com/twosigma/webtau/junit5/WebtauJunitExtension.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2018 TWO SIGMA OPEN SOURCE, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.twosigma.webtau.junit5;
+
+import com.twosigma.webtau.junit5.report.JavaBasedTest;
+import com.twosigma.webtau.junit5.report.ReportJavaTestEntries;
+import com.twosigma.webtau.report.HtmlReportGenerator;
+import com.twosigma.webtau.report.ReportTestEntry;
+import com.twosigma.webtau.reporter.StepReporters;
+import com.twosigma.webtau.reporter.TestResultPayloadExtractors;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestExecutionExceptionHandler;
+
+public class WebtauJunitExtension implements
+        BeforeEachCallback,
+        AfterEachCallback,
+        TestExecutionExceptionHandler {
+
+    private static final ExtensionContext.Namespace NAMESPACE = ExtensionContext.Namespace.create("webtau");
+    private static final String TEST_KEY = "test";
+
+    @Override
+    public void beforeEach(ExtensionContext extensionContext) throws Exception {
+        JavaBasedTest test = new JavaBasedTest(
+                extensionContext.getUniqueId(),
+                fullTestName(extensionContext));
+
+        StepReporters.add(test);
+        extensionContext.getStore(NAMESPACE).put(TEST_KEY, test);
+    }
+
+    @Override
+    public void afterEach(ExtensionContext extensionContext) throws Exception {
+        JavaBasedTest test = retrieveTest(extensionContext);
+
+        StepReporters.remove(test);
+
+        ReportTestEntry reportTestEntry = test.getReportTestEntry();
+        reportTestEntry.setRan(true);
+        ReportJavaTestEntries.add(reportTestEntry);
+
+        TestResultPayloadExtractors.extract(reportTestEntry.getSteps().stream())
+                .forEach(reportTestEntry::addTestResultPayload);
+
+        ShutdownHook.INSTANCE.noOp();
```

Comment:
Not sure I understand the benefit of this noOp invocation.  Is it just to force registration of the shutdown hook?  Do we need to actually have our ShutdownHook class to do that?

Reply:
I basically want to register hook only once, and only when any webtau test was called first. Wanted to avoid fields set and checks.

Reply:
Could you not register it in `WebtauJunitExtension`'s static initialiser block?  Not sure which of these options I dislike the least 😄

Reply:
I think in this case it will register hook even if the actual test is not called. Since Extension will be referred and therefore initialized during test files scan. I may be wrong.

Reply:
Nope, you're quite right.